docs/ADR-3-Dependencies-version-constraints-in-cabal-file.md

```diff
@@ -0,0 +1,87 @@
+Let's consider an example where multiple package *components* (e.g. libraries, internal libraries, test-suites,
+executables) form a more complex dependency graphs like the following one:
+```mermaid
+graph LR
+ B --> A
+ C --> A
+ D --> B
+ D --> C
+```
+Considering an example case, where `B` and `C` but not `A`, have the same dependencies, we need to encode version ranges twice.
+This results in duplicated entries in `.cabal` file, with duplicated version bounds.
+
+# Considered options
+
+1. **Using explicit common stanzas, one per dependency.** See for example: https://github.com/input-output-hk/hedgehog-extras/blob/e7f3c9ff967ed6f3b4c8c17013a4e0c2f541e053/hedgehog-extras.cabal#L19
+    * **Advantages**
+      * Fine grained control over what dependency is included where, with the single place which defines
+        version bounds.
+    * **Disadvantages**
+      * A lot of boilerplate when defining new dependencies.
+
+1. **Moving common dependencies to `common` stanzas, multiple dependencies in one stanza**.
+  This allows for grouping of dependencies between components in order specify version bound only once.
+  For example the same `common` stanza could be reused in `B` and `C` in order to define version bounds in a
+  single place.
+    * **Advantages**
+      * Grouping multiple dependencies together introduces less boilerplate in comparison to the
+        *one-stanza-per-one-dependency* option.
+    * **Disadvantages**
+      * It may be hard to define logical grouping for few unrelated dependencies.
+      * In the worst case scenario we'll end up with *one-stanza-per-one-dependency* definition of
+        dependencies, when those will be used in different components.
+
+1. **Declaring dependencies' bounds everywhere**.
+  In the provided example, the version bounds of common dependencies of `A` and `B` should be both defined in
+  `A` and `B`.
+    * **Advantages**
+      * Makes it easy to remember how to update dependency version everywhere.
+    * **Disadvantages**
+      * It's not hard to overlook one spot when a person updating dependency version is unfamiliar with the
+        rule.
+
+1. **Declaring dependencies' bounds once, as close as possible to the dependency tree root.**
+  In the provided example, the version bounds of common dependencies of `A` and `B` should be only defined in
+  `A`.
+  If dependencies are both present in `B` & `C`, the version bounds should be duplicated in both components.
+    * **Advantages**
+      * Slight de-duplication of version bounds definitions in comparison to duplicated version bounds
+      * Does not require introducing `common` stanzas
+    * **Disadvantages**
+      * In the `B` & `C` case, there will be a duplication of version bounds.
+      * Inconsistency in version bounds declaration because of `B` & `C` case.
```

Comment:
Disadvantage: If there are multiple roots then, then require all those roots would have to have bounds.  If one is accidentally missed it is possible for those bounds to be ignored if only roots without bounds are in the cabal plan, which could lead to the wrong version of the package build built.